futures-util/src/join.rs

```diff
@@ -0,0 +1,91 @@
+//! The `join` macro.
+
+use std::mem::{self, PinMut};
+use futures_core::{task, Future, Poll};
+
+#[macro_export]
+macro_rules! join {
+    ($($fut:ident),*) => { {
+        $(
+            let mut $fut = $crate::MaybeDone::Future($fut);
+            pin_mut!($fut);
+        )*
+        loop {
+            let mut all_done = true;
+            $(
+                if let ::core::task::Poll::Pending = poll!($fut.reborrow()) {
+                    all_done = false;
+                }
+            )*
+            if all_done {
+                break;
+            } else {
+                pending!();
+            }
+        }
+
+        ($(
+            $fut.reborrow().take_output().unwrap(),
+        )*)
+    } }
+}
+
+/// `MaybeDone`, a future that may have completed.
+#[doc(hidden)]
+#[derive(Debug)]
+pub enum MaybeDone<A: Future> {
```

Comment:
We should find a nice place for utility futures like this. `FutureOption`, `ReadyFuture`, `PendingOnce` etc. could all live there.